File: src/main/java/com/crud_javatokotiln/service/OrderService.java
package com.crud_javatokotiln.service;


import com.crud_javatokotiln.dto.OrderDto;
import com.crud_javatokotiln.entity.Order;
import com.crud_javatokotiln.entity.Product;
import com.crud_javatokotiln.entity.User;
import com.crud_javatokotiln.exception.NotFoundException;
import com.crud_javatokotiln.repository.OrderRepository;
import com.crud_javatokotiln.repository.ProductRepository;
import com.crud_javatokotiln.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class OrderService {

    private final OrderRepository orderRepository;
    private final UserRepository userRepository;
    private final ProductRepository productRepository;

    public List<Order> findAll() {
        return orderRepository.findAll();
    }

    public Order findById(Long id) {
        return orderRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Order not found with id: " + id));
    }

    @Transactional
    public Order create(OrderDto orderDto) {
        User user = userRepository.findById(orderDto.getUserId())
                .orElseThrow(() -> new NotFoundException("User not found with id: " + orderDto.getUserId()));
        Product product = productRepository.findById(orderDto.getProductId())
                .orElseThrow(() -> new NotFoundException("Product not found with id: " + orderDto.getProductId()));
        Order order = new Order(user, product, orderDto.getQuantity());
        return orderRepository.save(order);
    }

    @Transactional
    public Order update(Long id, OrderDto orderDto) {
        Order order = findById(id);
        User user = userRepository.findById(orderDto.getUserId())
                .orElseThrow(() -> new NotFoundException("User not found with id: " + orderDto.getUserId()));
        Product product = productRepository.findById(orderDto.getProductId())
                .orElseThrow(() -> new NotFoundException("Product not found with id: " + orderDto.getProductId()));
        order.setUser(user);
        order.setProduct(product);
        order.setQuantity(orderDto.getQuantity());
        return orderRepository.save(order);
    }

    @Transactional
    public void delete(Long id) {
        orderRepository.deleteById(id);
    }

    // 추가 기능: 주문 총액 계산 (수량 * 상품 가격)
    public double calculateTotal(Long orderId) {
        Order order = findById(orderId);
        return order.getQuantity() * order.getProduct().getPrice();
    }
}

File: src/main/java/com/crud_javatokotiln/service/OrderService.kt
package com.crud_javatokotiln.service

import com.crud_javatokotiln.dto.OrderDto
import com.crud_javatokotiln.entity.Order
import com.crud_javatokotiln.exception.NotFoundException
import com.crud_javatokotiln.repository.OrderRepository
import com.crud_javatokotiln.repository.ProductRepository
import com.crud_javatokotiln.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
open class OrderService(
    private val orderRepository: OrderRepository,
    private val userRepository: UserRepository,
    private val productRepository: ProductRepository
) {

    fun findAll(): List<Order> = orderRepository.findAll()

    fun findById(id: Long): Order =
        orderRepository.findById(id)
            .orElseThrow { NotFoundException("Order not found with id: $id") }

    @Transactional
    open fun create(orderDto: OrderDto): Order {
        val user = userRepository.findById(orderDto.userId)
            .orElseThrow { NotFoundException("User not found with id: ${orderDto.userId}") }!!
        val product = productRepository.findById(orderDto.productId)
            .orElseThrow { NotFoundException("Product not found with id: ${orderDto.productId}") }!!
        val order = Order(user, product, orderDto.quantity)
        return orderRepository.save(order)
    }

    @Transactional
    open fun update(id: Long, orderDto: OrderDto): Order {
        val order = findById(id)
        val user = userRepository.findById(orderDto.userId)
            .orElseThrow { NotFoundException("User not found with id: ${orderDto.userId}") }
        val product = productRepository.findById(orderDto.productId)
            .orElseThrow { NotFoundException("Product not found with id: ${orderDto.productId}") }
        order.user = user
        order.product = product
        order.quantity = orderDto.quantity
        return orderRepository.save(order)
    }

    @Transactional
    open fun delete(id: Long) {
        orderRepository.delete(findById(id))
    }


    fun calculateTotal(orderId: Long): Double {
        val order = findById(orderId)
        val product = order.product ?: throw NotFoundException("Product not found for order: $orderId")
        return order.quantity * product.price
    }

}
